src/lib/engine/factBank.js

```diff
@@ -0,0 +1,53 @@
+import generateRandomReadableId from './metric/generate_random_readable_id'
+import * as _guesstimate from './guesstimate'
+
+const byVariableName = name => f => f.variable_name === name
+const namedLike = partial => f => f.variable_name.startsWith(partial)
+const search = (partial, list) => _.isEmpty(partial) || !list ? '' : (_.get(list.filter(namedLike(partial))[0], 'variable_name') || '')
+
+const findChildrenBySelector = (facts, selector) => _.isEmpty(selector) ? facts : findBySelector(facts, selector).children
+export function selectorSearch(selector, facts) {
+  const partial = selector.pop()
+  const suggestion = search(partial, findChildrenBySelector(facts, selector))
+  return {partial, suggestion}
+}
+
+export const idFrom = selector => selector.join('.')
+export const toMetric = (selector, takenReadableIds = []) => ({id: idFrom(selector), readableId: generateRandomReadableId(takenReadableIds)})
+export const toGuesstimate = (selector, {expression}) => ({metric: idFrom(selector), input: expression})
+export const toSimulation = (selector, {values}) => ({metric: idFrom(selector), sample: {values}})
+
+export function findBySelector(facts, selector, currFact = {}) {
+  if (_.isEmpty(selector)) { return currFact }
+  if (_.isEmpty(facts)) { return {} }
+  const fact = facts.find(byVariableName(selector[0]))
+  if (!fact) { return {} }
+  return findBySelector(fact.children, selector.slice(1), fact)
+}
+
+// Currently only supports globals
+const resolveToSelector = handle => handle.slice(1).split('.')
+export function addFactsToSpaceGraph({metrics, guesstimates, simulations}, globals, organizationId) {
```

Comment:
Add comments. The forEach method internals being split up seemed like it would require the split off method to take a lot of params and return a lot of params, so didn't seem wise.